ADR/0035-artifact-streams.md

```diff
@@ -0,0 +1,150 @@
+own relevant internal data model.
+
+### ArtifactConfig
+
+An `ArtifactConfig` describes the metadata which is used to define an `Artifact`, an `ArtifactStream`, and
+how `Artifact`s produced update that stream.
+
+`Artifact`s do not exist in isolation to other `Artifact`s in the system, however, there are both explicit
+and implicit relationships that exist between them, for example:
+
+- (implicit) Multiple `ArtifactConfig`s can be created for the same repository and branch combinations. These
+  configurations would be allowed as long as the name of the produced artifact is unique. This supports use
+  cases like `ko` where multiple `Artifact`s would be produced from a single build
+- (explicit) A single `ArtifactConfig` can describe the configuration for producing the "same" `Artifact"
+  from multiple branches in the same git repository. As an `ArtifactConfig` also controls the creation of
+  an `ArtifactStream`, `Artifact`s produced on any one of the branches would update the corresponding
+  `ArtifactStream`.
+- (explicit) A single repository and branch may produce different `Artifact`s under different conditions.
+  When artifacts are produced as part of a pull request, for example, the controller will still need to
+  generate `Artifact` resources, but these should not be able to update any `ArtifactStream`s for the 
+  on-push build events
+
+### Artifact
+
+An `Artifact` represents a single object (binary or non-binary artifacts) produced as part of a software
+supply chain. It not only defines the storage location of the object itself but also the location of any
+additional supporting metadata like produced SBOMs, provenance, and signatures (and the public gpg keys).
+
+Since the concept of an `Artifact` exists outside of any other controller, the object may either be manually
+created based on known artifacts or a constructor may be leveraged to watch for specific actions (i.e. within
+the kubernetes cluster or new artifacts being pushed to an external location) in order to produce the resulting
+`Artifact`. An example of this constructor can resemble the `TestSubjectConstructor` as proposed in
+[Upstreaming integration-service].
+
+`Artifact`s can also store references to the process which created the object (i.e. PipelineRun). As soon as a
+PipelineRun is created, an `Artifact` should be created to enable visibility into the ongoing process if
+an `ArtifactStream` is used for tracking or visualization.
+
+### ArtifactStream
+
+All `Artifact`s have an owner reference to a single `ArtifactStream`. The `ArtifactStream` keeps track of the latest
+produced `Artifact` that has been created owned by that stream (i.e. similar to a floating tag reference).
+
+A validation webhook will prevent an `Artifact` from being created unless its configuration matches that of its owning
+`ArtifactStream`.
+
+### ArtifactCollection
+
+An `ArtifactCollection` is a set of `Artifact`s that are grouped together. An `ArtifactCollection` can include
```

Comment:
What would a typical grouping of `ArtifactCollections` contain, in terms of types of `Artifact`s?


And could we achieve a similar level of grouping with labels on the resourecs, rather than creating another resource type to group them? Maybe something like:
```
labels:
   artifactCollection: <some-collection>
```